cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	//}
+
+	// create service instance
+	s.Info("Creating new service instance", "name", s.GetServiceName(ServiceInstance))
+	serviceInstance, _, err := s.ResourceClient.CreateResourceInstance(&resourcecontrollerv2.CreateResourceInstanceOptions{
+		Name:           s.GetServiceName(ServiceInstance),
+		Target:         s.Zone(),
+		ResourceGroup:  &resourceGroupID,
+		ResourcePlanID: pointer.String(powerVSResourcePlanID),
+	})
+	if err != nil {
+		return nil, err
+	}
+	s.Info("Created new service instance")
+	return serviceInstance, nil
+}
+
+// ReconcileNetwork reconciles network.
+func (s *PowerVSClusterScope) ReconcileNetwork() error {
+	// if DHCP server id is set means the server is already created
+	if s.GetDHCPServerID() != nil {
+		s.Info("DHCP server id is set")
+		err := s.checkDHCPServerStatus()
+		if err != nil {
+			return err
+		}
+		return nil
+		//	TODO(karthik-k-n): If needed set dhcp status here
+	}
+	// check network exist in cloud
+	networkID, err := s.checkNetwork()
+	if err != nil {
+		return err
+	}
+	if networkID != "" {
+		s.Info("Found network", "id", networkID)
+		s.SetStatus(Network, infrav1beta2.ResourceReference{ID: &networkID, ControllerCreated: pointer.Bool(false)})
+		return nil
+	}
+
+	s.Info("Creating DHCP server")
+	dhcpServer, err := s.createDHCPServer()
+	if err != nil {
+		s.Error(err, "Error creating DHCP server")
+		return err
+	}
+	if dhcpServer != nil {
+		s.Info("Created DHCP Server", "id", *dhcpServer)
+		s.SetStatus(DHCPServer, infrav1beta2.ResourceReference{ID: dhcpServer, ControllerCreated: pointer.Bool(true)})
+		return nil
+	}
+
+	err = s.checkDHCPServerStatus()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// checkNetwork checks the network exist in cloud.
+func (s *PowerVSClusterScope) checkNetwork() (string, error) {
+	networkID, err := s.getNetwork()
+	if err != nil {
+		s.Error(err, "failed to get network")
+		return "", err
+	}
+	if networkID == nil {
+		s.Info("Not able to find network", "network", s.IBMPowerVSCluster.Spec.Network)
+		return "", nil
+	}
+	return *networkID, nil
+}
+
+func (s *PowerVSClusterScope) getNetwork() (*string, error) {
+	if s.IBMPowerVSCluster.Spec.Network.ID != nil {
+		network, err := s.IBMPowerVSClient.GetNetworkByID(*s.IBMPowerVSCluster.Spec.Network.ID)
+		if err != nil {
+			return nil, err
+		}
+		return network.NetworkID, nil
+	}
+	network, err := s.IBMPowerVSClient.GetNetworkByName(*s.GetServiceName(Network))
+	if err != nil {
+		return nil, err
+	}
+	if network == nil {
+		s.Info("network does not exist", "name", *s.GetServiceName(Network))
+		return nil, nil
+	}
+	return network.NetworkID, nil
+	//TODO: Support regular expression
+}
+
+// checkDHCPServerStatus checks the DHCP server status.
+func (s *PowerVSClusterScope) checkDHCPServerStatus() error {
+	dhcpID := *s.GetDHCPServerID()
+	dhcpServer, err := s.IBMPowerVSClient.GetDHCPServer(dhcpID)
+	if err != nil {
+		return err
+	}
+	if dhcpServer == nil {
```

Comment:
we can remove this check, dhcpclient.get returns either err or dhcpserver obj. If err is nil, dhcpserver obj won't be nil.